File: selenium/run_e2e.js
#!/usr/bin/env node
const { spawnSync } = require("child_process");
const fs = require("fs");
const path = require("path");

function parseArgs(argv) {
  const out = { driver: process.env.E2E_DRIVER || "simulated" };
  for (let i = 0; i < argv.length; i += 1) {
    if (argv[i] === "--driver" && argv[i + 1]) {
      out.driver = argv[i + 1];
      i += 1;
    }
  }
  return out;
}

function main() {
  const args = parseArgs(process.argv.slice(2));
  process.env.E2E_DRIVER = args.driver;

  console.log(`Running Selenium E2E with driver mode: ${args.driver}`);
  const mochaBin = path.join(__dirname, "node_modules", "mocha", "bin", "mocha.js");
  const result = spawnSync(
    process.execPath,
    [mochaBin, "tests/**/*.test.js", "--timeout", "60000"],
    {
      cwd: __dirname,
      stdio: "inherit",
      env: process.env,
    }
  );

  const report = path.join(
    __dirname,
    "reports",
    "Selenium_E2E_Excel_Analysis_Report.xlsx"
  );
  if (!fs.existsSync(report)) {
    console.error("WARNING: Excel report was not generated");
    process.exit(result.status === 0 ? 1 : result.status || 1);
  }
  console.log(`Excel analysis report: ${report}`);
  process.exit(result.status || 0);
}

main();

File: appium/run_e2e.py
"""Run Appium E2E suite and ensure Excel analysis report is produced."""

from __future__ import annotations

import argparse
import os
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent


def main() -> int:
    parser = argparse.ArgumentParser(description="Smart Furnace Appium E2E runner")
    parser.add_argument(
        "--driver",
        choices=["simulated", "appium"],
        default=os.getenv("E2E_DRIVER", "simulated"),
    )
    args = parser.parse_args()
    os.environ["E2E_DRIVER"] = args.driver

    print(f"Running Appium E2E with driver mode: {args.driver}")
    result = subprocess.run(
        [sys.executable, "-m", "pytest", "tests", "-v", "--tb=short"],
        cwd=str(ROOT),
        check=False,
    )
    report = ROOT / "reports" / "Appium_E2E_Excel_Analysis_Report.xlsx"
    if report.exists():
        print(f"Excel analysis report: {report}")
    else:
        print("WARNING: Excel report was not generated")
        return 1 if result.returncode == 0 else result.returncode
    return result.returncode


if __name__ == "__main__":
    raise SystemExit(main())
